app/assets/javascripts/oxalis/model/bucket_data_handling/pullqueue.js

```diff
@@ -93,56 +93,70 @@ class PullQueue {
+          const higherAddress = zoomedAddressToAnotherZoomStep(
+            bucketAddress,
+            resolutions,
+            zoomStep + 1,
+          );
+
+          const resolutionsFactors = getResolutionsFactors(
+            resolutions[zoomStep + 1],
+            resolutions[zoomStep],
+          );
+          const higherBucket = this.cube.getOrCreateBucket(higherAddress);
+          if (higherBucket.type === "data") {
+            higherBucket.downsampleFromLowerBucket(
+              bucket,
+              resolutionsFactors,
+              layerInfo.category === "segmentation",
+            );
+          }
+        }
+      }
+    };
+
+    const { renderMissingDataBlack } = Store.getState().datasetConfiguration;
+
     try {
-      const responseBuffer = await requestFromStore(layerInfo, batch);
-      let bucketData;
+      const { buffer: responseBuffer, missingBuckets } = await requestFromStore(layerInfo, batch);
       this.connectionInfo.log(
         this.layerName,
         roundTripBeginTime,
         batch.length,
         responseBuffer.length,
       );
 
-      let offset = 0;
       const resolutions = getResolutions(dataset);
-      for (const bucketAddress of batch) {
-        const zoomStep = bucketAddress[3];
-        if (zoomStep > this.cube.MAX_UNSAMPLED_ZOOM_STEP) {
+      let offset = 0;
+      for (const [index, bucketAddress] of batch.entries()) {
+        const isMissing = missingBuckets.indexOf(index) > -1;
+        const bucket = this.cube.getBucket(bucketAddress);
+        if (bucket.type !== "data") {
           continue;
         }
-        bucketData = responseBuffer.subarray(offset, (offset += this.cube.BUCKET_LENGTH));
-        const bucket = this.cube.getBucket(bucketAddress);
-        this.cube.boundingBox.removeOutsideArea(bucket, bucketAddress, bucketData);
-        this.maybeWhitenEmptyBucket(bucketData);
-        if (bucket.type === "data") {
-          bucket.receiveData(bucketData);
-          if (zoomStep === this.cube.MAX_UNSAMPLED_ZOOM_STEP) {
-            const higherAddress = zoomedAddressToAnotherZoomStep(
-              bucketAddress,
-              resolutions,
-              zoomStep + 1,
-            );
-
-            const resolutionsFactors = getResolutionsFactors(
-              resolutions[zoomStep + 1],
-              resolutions[zoomStep],
-            );
-            const higherBucket = this.cube.getOrCreateBucket(higherAddress);
-            if (higherBucket.type === "data") {
-              higherBucket.downsampleFromLowerBucket(
-                bucket,
-                resolutionsFactors,
-                layerInfo.category === "segmentation",
-              );
-            }
-          }
+        if (isMissing && !renderMissingDataBlack) {
+          bucket.pullFailed(true);
+        } else {
+          const bucketData = isMissing
+            ? new Uint8Array(bucket.BUCKET_LENGTH)
+            : responseBuffer.subarray(offset, (offset += this.cube.BUCKET_LENGTH));
```

Comment:
Is there a difference between `bucket.BUCKET_LENGTH` (previous line) and `this.cube.BUCKET_LENGTH`? If not, I would use only one of the two variants.